wallets/server.js

```diff
@@ -11,120 +11,231 @@ import * as blink from 'wallets/blink/server'
+          createInvoice({
+            msats: toPositiveNumber(innerMsats),
+            description,
+            descriptionHash,
+            expiry
+          }, walletFull[walletField]), 10_000)
+        await checkInvoice(models, wallet, invoice, innerMsats)
+        if (!lnd) throw new Error('no lnd provided for wrapping')
+        const { invoice: wrappedInvoice, maxFee } = await wrapInvoice(invoice,
+          {
+            msats: wrappedMsats,
+            description: !descriptionHash ? description : undefined,
+            descriptionHash
+          }, { lnd }, sybilFeePercent)
+        await checkInvoice(models, wallet, wrappedInvoice.request, wrappedMsats)
+        return {
+          innerInvoice: invoice,
+          invoice: wrappedInvoice.request,
+          wallet,
+          maxFee,
+          isWrapped: true,
+          isHodl: true,
           msats,
-          description: wallet.user.hideInvoiceDesc ? undefined : description,
-          descriptionHash,
-          expiry
-        }, walletFull[walletField]), 10_000)
-
-      const bolt11 = await parsePaymentRequest({ request: invoice })
-      if (BigInt(bolt11.mtokens) !== BigInt(msats)) {
-        if (BigInt(bolt11.mtokens) > BigInt(msats)) {
-          throw new Error(`invoice is for an amount greater than requested ${bolt11.mtokens} > ${msats}`)
-        }
-        if (BigInt(bolt11.mtokens) === 0n) {
-          throw new Error('invoice is for 0 msats')
+          description,
+          retriable: i < wallets.length - 1,
+          attempt
         }
-        if (BigInt(msats) - BigInt(bolt11.mtokens) >= 1000n) {
-          throw new Error(`invoice has a different satoshi amount ${bolt11.mtokens} !== ${msats}`)
-        }
-
-        await addWalletLog({
+      } else {
+      // return a regular invoice straight from the attached wallet
+        const invoice = await withTimeout(
+          createInvoice({
+            msats,
+            description,
+            descriptionHash,
+            expiry
+          }, walletFull[walletField]), 10_000)
+        await checkInvoice(models, wallet, invoice, msats)
+
+        return {
+          invoice,
           wallet,
-          level: 'INFO',
-          message: `wallet does not support msats so we floored ${msats} msats to nearest sat ${BigInt(bolt11.mtokens)} msats`
-        }, { models })
+          isWrapped: false,
+          isHodl: false,
+          msats,
+          description,
+          retriable: i < wallets.length - 1,
+          attempt
+        }
       }
-
-      return { invoice, wallet }
     } catch (error) {
-      console.error(error)
+      console.error('error checking wallet:', error)
       await addWalletLog({
         wallet,
         level: 'ERROR',
         message: `creating invoice for ${description ?? ''} failed: ` + error
       }, { models })
     }
   }
-
   throw new Error('no wallet available')
 }
+
+export const addWalletLog = async ({ wallet, level, message }, { models }) => {
```

Comment:
this was moved here to avoid cyclic dependencies with this file and the resolver

Reply:
this function was removed in #994